[SPARK-12073] [Streaming] backpressure rate controller consumes events preferentially from lagg…

JasonMWhite opened the pull request:
…ing partitions

I'm pretty sure this is the reason we couldn't easily recover from an unbalanced Kafka partition under heavy load when using backpressure.

`maxMessagesPerPartition` calculates an appropriate limit for the message rate from all partitions, and then divides by the number of partitions to determine how many messages to retrieve per partition. The problem with this approach is that when one partition is behind by millions of records (due to random Kafka issues), but the rate estimator calculates only 100k total messages can be retrieved, each partition (out of say 32) only retrieves max 100k/32=3125 messages.

This PR (still needing a test) determines a per-partition desired message count by using the current lag for each partition to preferentially weight the total message limit among the partitions. In this situation, if each partition gets 1k messages, but 1 partition starts 1M behind, then the total number of messages to retrieve is (32 \* 1k + 1M) = 1032000 messages, of which the one partition needs 1001000. So, it gets (1001000 / 1032000) = 97% of the 100k messages, and the other 31 partitions share the remaining 3%.

Assuming all of 100k the messages are retrieved and processed within the batch window, the rate calculator will increase the number of messages to retrieve in the next batch, until it reaches a new stable point or the backlog is finished processed.

We're going to try deploying this internally at Shopify to see if this resolves our issue.

@tdas @koeninger @holdenk

Commit:
backpressure rate controller consumes events preferentially from lagging partitions

AmplabJenkins commented:
Can one of the admins verify this patch?

koeninger commented on external/kafka/src/test/scala/org/apache/spark/streaming/kafka/DirectKafkaStreamSuite.scala:
```diff
@@ -364,8 +365,8 @@ class DirectKafkaStreamSuite
-    val messageKeys = (1 to 200).map(_.toString)
-    val messages = messageKeys.map((_, 1)).toMap
+    val messages = Map("foo" -> 200)
+    kafkaTestUtils.sendMessages(topic, messages)
```
Why was sendMessages moved to here?

JasonMWhite replied:
The kafka messages don't actually need to be sent three times, just once is sufficent for all tests. When I send "foo" 200 times in one batch, they all go to the same partition. However, when I do this 3 times (for each of 100, 50, 20), the batches of 200 go to a random partition each time. I suspect something in how the test kafka cluster does the partitioning.

I was usually getting 200 on 1 partitions, and 400 on the other 2. I was explicitly changing the test case to "all the messages are on one partition" since I can't control the split deterministically.

holdenk replied:
While I haven't done much work with Kafka, it seems like we could maybe explicitly specify the partitioner for producer in the producerConfiguration to be round robin if we wanted to (although that requires some custom code from what I can tell) or rotating the partition key.

koeninger replied:
Not sure if you're just talking about for testing, or for the main code.
For production use the complication is that how messages are partitioned
into kafka is configurable at the time you're producing the messages, so
configuration of spark partitioner would have to match.

On Thu, Dec 10, 2015 at 2:14 PM, Holden Karau <notifications@github.com>
wrote:

> In
> external/kafka/src/test/scala/org/apache/spark/streaming/kafka/DirectKafkaStreamSuite.scala
> <https://github.com/apache/spark/pull/10089#discussion_r47278574>:
>
> > @@ -364,8 +365,8 @@ class DirectKafkaStreamSuite
> >
> >      val batchIntervalMilliseconds = 100
> >      val estimator = new ConstantEstimator(100)
> > -    val messageKeys = (1 to 200).map(_.toString)
> > -    val messages = messageKeys.map((_, 1)).toMap
> > +    val messages = Map("foo" -> 200)
> > +    kafkaTestUtils.sendMessages(topic, messages)
>
> While I haven't done much work with Kafka, it seems like we could maybe
> explicitly specify the partitioner for producer in the
> producerConfiguration to be round robin if we wanted to (although that
> requires some custom code from what I can tell) or rotating the partition
> key.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/10089/files#r47278574>.
>

holdenk replied:
Oh yah only for testing - this was in response to switching the test to all messages on a single partition (which seemed limiting for testing code which changes us to handling each partitions back pressure instead of a single global).

JasonMWhite replied:
In the original code, as I understand it, since the Kafka test setup wasn't cleared/reinitialized between testing rounds, only the first batch of 200 messages, produced in the first of 3 test rounds, was ever consumed. The other testing rounds produced messages that were never used. My changes aside, I think moving the test message generation outside of the individual test rounds makes the most sense.

This failure scenario depended on imbalanced Kafka partitions, would you prefer to see tests on both a balanced and an imbalanced scenario?

koeninger commented on external/kafka/src/test/scala/org/apache/spark/streaming/kafka/DirectKafkaStreamSuite.scala:
```diff
@@ -36,9 +36,10 @@ import org.scalatest.concurrent.Eventually
 import org.apache.spark.{Logging, SparkConf, SparkContext, SparkFunSuite}
 import org.apache.spark.rdd.RDD
-import org.apache.spark.streaming.{Milliseconds, StreamingContext, Time}
+import org.apache.spark.streaming.{kafka, Milliseconds, StreamingContext, Time}
```
This is just an IDE unnecessarily mucking with imports, right?

JasonMWhite replied:
yes, sorry about that, I'll fix that up

koeninger commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala:
```diff
@@ -89,23 +89,29 @@ class DirectKafkaInputDStream[
+        val totalLag = lagPerPartition.values.sum.toFloat
+
+        lagPerPartition.map { case (tp, lag) =>
+          tp -> Math.round(lag / Math.max(totalLag, 1) * rate)
```
What happens when you only have 1 partition with lag > 0 ?
I think this needs tests for boundary conditions

koeninger replied:
I also think you should account for what happens when someone has set
spark.streaming.kafka.maxRatePerPartition
in addition to enabling backpressure

JasonMWhite replied:
Good point!

JasonMWhite replied:
The current test puts all events on a single partition, while the other partition has no events, and therefore no lag. I could add a test for the boundary condition of no lag on any partitions, also one for where the current offset is less than the latest offset (can happen in unclean election situations).

Definitely needs to respect `maxRatePerPartition` even when backpressure is working, so that needs some adjustment.

koeninger commented:
This generally looks sensible to me, would like to see if it solves your issue first. Thanks for working on it.

JasonMWhite commented:
This patch solved our skew problem. Below is a 15-minute snapshot of our lag earlier this week, showing a single partition getting slowly worse. It would get to about 8 million messages behind overnight.

https://monosnap.com/file/VG0rXFZn05bKIIDLFIHOxfNOWKOxy5

And the most recent 24 hour period, after this patch went live on our system.

https://monosnap.com/file/DMSzLP8lFKJ0DxHuiWcxZHKqVeJHWL

koeninger commented:
Sounds good. If you can add that maxRatePerPartition handling this would be
ready to go from my point of view
On Dec 5, 2015 3:59 PM, "Jason White" <notifications@github.com> wrote:

> This patch solved our skew problem. Below is a 15-minute snapshot of our
> lag earlier this week, showing a single partition getting slowly worse. It
> would get to about 8 million messages behind overnight.
>
> https://monosnap.com/file/VG0rXFZn05bKIIDLFIHOxfNOWKOxy5
>
> And the most recent 24 hour period, after this patch went live on our
> system.
>
> https://monosnap.com/file/DMSzLP8lFKJ0DxHuiWcxZHKqVeJHWL
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/10089#issuecomment-162250749>.
>

JasonMWhite commented:
`maxRatePerPartition` now respects the limit set by `maxRateLimitPerPartition`, if it is set. Let me know if you think any additional tests are needed.

koeninger commented:
LGTM

JasonMWhite commented:
Could someone verify the patch please? @tdas perhaps?

mrszg commented:
Original code (before this patch) has serious error - it doesn't respect maxRateLimitPerPartition in case when backpressure rate is smaller than number of partitions. In such case effectiveRateLimitPerPartition equals 0 (because limit / numPartitions == 0) and in consequence all messages from topic are consumed at once. According my tests this patch solves this error.

zsxwing commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47665/
Test FAILed.

JasonMWhite commented:
Is this a CI failure? I don't have access rights to see what happened. I'd like to get this ready to merge, could someone give me a hand?

koeninger commented:
The console output seems like it's available without logging in, and looks like a jenkins issue rather than an actual test failure:

GitHub pull request #10089 of commit b58d51767f0370c65fa65dfb15416b75fa914d05 automatically merged.
[EnvInject] - Loading node environment variables.
Building remotely on amp-jenkins-worker-07 (centos spark-test) in workspace /home/jenkins/workspace/SparkPullRequestBuilder
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/apache/spark.git # timeout=10
Fetching upstream changes from https://github.com/apache/spark.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/apache/spark.git +refs/pull/10089/*:refs/remotes/origin/pr/10089/* # timeout=15
Build was aborted
Aborted by anonymous
ERROR: Step ?Archive the artifacts? failed: no workspace for SparkPullRequestBuilder #47665
ERROR: Step ?Publish JUnit test result report? failed: no workspace for SparkPullRequestBuilder #47665
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47665/
Test FAILed.
ERROR: amp-jenkins-worker-07 is offline; cannot locate JDK 7u60

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47935/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #47938 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47938/consoleFull)** for PR 10089 at commit [`b58d517`](https://github.com/apache/spark/commit/b58d51767f0370c65fa65dfb15416b75fa914d05).

SparkQA commented:
**[Test build #47938 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47938/consoleFull)** for PR 10089 at commit [`b58d517`](https://github.com/apache/spark/commit/b58d51767f0370c65fa65dfb15416b75fa914d05).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47938/
Test FAILed.

koeninger commented:
This looked like a good patch, did it just fall through the cracks?  The mima test failure was probably just due to the change in signature of KafkaTestUtils.createTopic, it should be ok to add an exclude for that since it's a testing class.

Jason if you don't have time to deal with it let me know + I can fix it

JasonMWhite commented:
I'm not sure how to handle the mima test failure, could you point me to where to add the exclude? I'll rebase and add the exception.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #50983 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50983/consoleFull)** for PR 10089 at commit [`b58d517`](https://github.com/apache/spark/commit/b58d51767f0370c65fa65dfb15416b75fa914d05).

SparkQA commented:
**[Test build #50983 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50983/consoleFull)** for PR 10089 at commit [`b58d517`](https://github.com/apache/spark/commit/b58d51767f0370c65fa65dfb15416b75fa914d05).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50983/
Test FAILed.

koeninger commented:
Sure, it's in project/MimaExcludes.scala.  You should be able to match up
the problem type and class in the error message from jenkins when adding an
appropriate ProblemFilters.exclude line.

On Tue, Feb 9, 2016 at 12:22 PM, Jason White <notifications@github.com>
wrote:

> I'm not sure how to handle the mima test failure, could you point me to
> where to add the exclude? I'll rebase and add the exception.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/10089#issuecomment-181990835>.
>

JasonMWhite commented:
Thanks!

Commit:
adding KafkaTestUtils.createTopic and DirectKafkaInputDStream.maxMessagesPerPartition to MimaExcludes

JasonMWhite commented:
@koeninger I've added the two error messages from Jenkins to the MimaExcludes, under the v2.0 section. There was one from the test function, but this PR also changes the signature of the protected method `maxMessagesPerPartition`.

koeninger commented:
I think that should be ok.

On Tue, Feb 16, 2016 at 4:29 PM, Jason White <notifications@github.com>
wrote:

> @koeninger <https://github.com/koeninger> I've added the two error
> messages from Jenkins to the MimaExcludes, under the v2.0 section. There
> was one from the test function, but this PR also changes the signature of
> the protected method maxMessagesPerPartition.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/10089#issuecomment-184900516>.
>

zsxwing commented:
ok to test

SparkQA commented:
**[Test build #51390 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51390/consoleFull)** for PR 10089 at commit [`f7ffd6f`](https://github.com/apache/spark/commit/f7ffd6f4b148ca56927e553970f464c907ced38a).

SparkQA commented:
**[Test build #51390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51390/consoleFull)** for PR 10089 at commit [`f7ffd6f`](https://github.com/apache/spark/commit/f7ffd6f4b148ca56927e553970f464c907ced38a).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51390/
Test FAILed.

JasonMWhite commented:
Sorry, did I put the MiMa exclusion in the wrong section of the file?

zsxwing commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala:
```diff
@@ -89,23 +89,31 @@ class DirectKafkaInputDStream[
+        val lagPerPartition = leaderOffsets.map { case (tp, lo) =>
+          tp -> Math.max(lo.offset - currentOffsets(tp), 0)
+        }
+        val totalLag = lagPerPartition.values.sum.toFloat
```
nit: If `lagPerPartition.values.sum` is 0, you can just return `None`. Then `Math.max(totalLag, 1)` can be changed to `totalLag`. E.g.,

```
val totalLag = lagPerPartition.values.sum
if (totalLag == 0) {
  return None
}
...
val backpressureRate = Math.round(lag.toFloat / totalLag * rate)
```

zsxwing commented:
@JasonMWhite could you add a unit test for `maxMessagesPerPartition`? Since it's protected, you should be able to test it using some faking data.

zsxwing commented:
> Sorry, did I put the MiMa exclusion in the wrong section of the file?

Then should be in 2.0 section.

Commit:
adding unit tests for maxMessagesPerPartition and fixing MimaExcludes

JasonMWhite commented:
I added unit tests for `maxMessagesPerPartition` to cover 3 edge cases that have been raised here:
- when backpressure is disabled, simply uses the `maxRatePerPartition`
- when there is no lag, returns `None`
- when backpressure calculations returns higher than `maxRatePerPartition`, respects `maxRatePerPartition`

The normal test, using the fake kafka test data, exercises `maxMessagePerPartition` in the normal case, i.e. backpressure is enabled, there is lag, backpressure is lower than the max rate per partition.

I've also moved the MimaExcludes to the 2.0 section.

SparkQA commented:
**[Test build #51451 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51451/consoleFull)** for PR 10089 at commit [`73e9ae3`](https://github.com/apache/spark/commit/73e9ae3ce7cc14c8ca50d8e385ad2604cfbf76f0).

zsxwing commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala:
```diff
@@ -89,23 +89,32 @@ class DirectKafkaInputDStream[
-  protected def maxMessagesPerPartition: Option[Long] = {
+
+  protected[streaming] def maxMessagesPerPartition(
+    offsets: Map[TopicAndPartition, Long]): Option[Map[TopicAndPartition, Long]] = {
```
nit: 4 spaces

JasonMWhite replied:
Thanks! I haven't quite got the hang on Scala alignment.

zsxwing commented on external/kafka/src/test/scala/org/apache/spark/streaming/kafka/DirectKafkaStreamSuite.scala:
```diff
@@ -353,10 +353,52 @@ class DirectKafkaStreamSuite
 
+  test("maxMessagesPerPartition with backpressure disabled") {
+    val topic = "maxMessagesPerPartition"
+    def getRateController(
```
`getRateController` is not necessary. You can just pass a rate controller to `getDirectKafkaStream` directly

JasonMWhite replied:
That is so much simpler.

zsxwing commented on external/kafka/src/test/scala/org/apache/spark/streaming/kafka/DirectKafkaStreamSuite.scala:
```diff
@@ -430,6 +471,32 @@ class DirectKafkaStreamSuite
+    val earliestOffsets = Map(TopicAndPartition(topic, 0) -> 0L, TopicAndPartition(topic, 1) -> 0L)
+    val messageHandler = (mmd: MessageAndMetadata[String, String]) => (mmd.key, mmd.message)
+    new DirectKafkaInputDStream[String, String, StringDecoder,
+                                StringDecoder, (String, String)](
```
nit: can be fit into one line

zsxwing commented on external/kafka/src/test/scala/org/apache/spark/streaming/kafka/DirectKafkaStreamSuite.scala:
```diff
@@ -430,6 +471,32 @@ class DirectKafkaStreamSuite
+    val messageHandler = (mmd: MessageAndMetadata[String, String]) => (mmd.key, mmd.message)
+    new DirectKafkaInputDStream[String, String, StringDecoder,
+                                StringDecoder, (String, String)](
+      ssc, Map[String, String](), earliestOffsets, messageHandler) {
```
nit: 4 spaces

SparkQA commented:
**[Test build #51451 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51451/consoleFull)** for PR 10089 at commit [`73e9ae3`](https://github.com/apache/spark/commit/73e9ae3ce7cc14c8ca50d8e385ad2604cfbf76f0).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51451/
Test FAILed.

Commit:
simplifying mock rate controller override, fixing alignment

JasonMWhite commented:
Changes addressed, but looks like it's causing PySpark unit tests to fail. Investigating...

JoshRosen commented:
```
py4j.Py4JException: Method createTopic([class java.lang.String]) does not exist
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:335)
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:344)
	at py4j.Gateway.invoke(Gateway.java:279)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:209)
	at java.lang.Thread.run(Thread.java:745)
```

From Java clients' POV, the signature of `createTopic()` in KafkaTestUtils has changed, so you'll need to explicitly provide the extra argument from the Python caller or will need to add back a single-argument constructor for backwards-compatibility.

SparkQA commented:
**[Test build #51457 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51457/consoleFull)** for PR 10089 at commit [`7a5dad3`](https://github.com/apache/spark/commit/7a5dad38e705c79a13200b23749e347cd782b02f).

Commit:
adding single-argument version of createTopic for backwards compatibility

JasonMWhite commented:
A single-argument version is easy enough, and good to support backwards-compatibility anyway. I haven't been able to get PySpark tests running locally yet, so I'm afraid that's a blind fix.

Commit:
fixing code alignment

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51459/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51463/
Test FAILed.

SparkQA commented:
**[Test build #51457 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51457/consoleFull)** for PR 10089 at commit [`7a5dad3`](https://github.com/apache/spark/commit/7a5dad38e705c79a13200b23749e347cd782b02f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51457/
Test FAILed.

JasonMWhite commented:
It looks like this PySpark unit test failure above was prior to my commit to add a backwards-compatible single-argument version as suggested. Could someone kick it again?

zsxwing commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala:
```diff
@@ -152,12 +152,15 @@ private[kafka] class KafkaTestUtils extends Logging {
   /** Create a Kafka topic and wait until it is propagated to the whole cluster */
-  def createTopic(topic: String): Unit = {
-    AdminUtils.createTopic(zkClient, topic, 1, 1)
+  def createTopic(topic: String, partitions: Int = 1): Unit = {
```
Could you remove the default value `1` since you added the overload version?

zsxwing commented:
retest this please

zsxwing commented on external/kafka/src/test/scala/org/apache/spark/streaming/kafka/DirectKafkaStreamSuite.scala:
```diff
@@ -430,6 +457,27 @@ class DirectKafkaStreamSuite
+    val batchIntervalMilliseconds = 100
+
+    val sparkConf = new SparkConf()
+      // Safe, even with streaming, because we're using the direct API.
```
Remove this comment. It's a bit misleading. Since the tests won't start the StreamingContext, the core number is irrelevant.

zsxwing commented:
@JasonMWhite thanks, looks great except some nits.

SparkQA commented:
**[Test build #51525 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51525/consoleFull)** for PR 10089 at commit [`f19f746`](https://github.com/apache/spark/commit/f19f7465c959505d7af56a375400b202e01865ef).

SparkQA commented:
**[Test build #51525 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51525/consoleFull)** for PR 10089 at commit [`f19f746`](https://github.com/apache/spark/commit/f19f7465c959505d7af56a375400b202e01865ef).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51525/
Test FAILed.

koeninger commented:
@JasonMWhite looks like this failed the "offset recovery" test in DirectKafkaStreamSuite.  Are you able to reproduce that test failure locally?

SparkQA commented:
**[Test build #52299 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52299/consoleFull)** for PR 10089 at commit [`0a78e8d`](https://github.com/apache/spark/commit/0a78e8d48a55df4a51e492451f73e8d95094bdba).

Commit:
removing unnecessary comment, default value

SparkQA commented:
**[Test build #52304 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52304/consoleFull)** for PR 10089 at commit [`a7a0877`](https://github.com/apache/spark/commit/a7a08771eb9e7fa9d15ff739c0da89fd193f8ed3).

JasonMWhite commented:
`DirectKafkaStreamSuite` passes all tests for me locally, but the test failure above appeared to be on an outdated sha.

Addressed @zsxwing's comments also. Waiting for the test build to complete.

SparkQA commented:
**[Test build #52299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52299/consoleFull)** for PR 10089 at commit [`0a78e8d`](https://github.com/apache/spark/commit/0a78e8d48a55df4a51e492451f73e8d95094bdba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52299/
Test PASSed.

SparkQA commented:
**[Test build #52304 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52304/consoleFull)** for PR 10089 at commit [`a7a0877`](https://github.com/apache/spark/commit/a7a08771eb9e7fa9d15ff739c0da89fd193f8ed3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52304/
Test PASSed.

JasonMWhite commented:
All comments addressed, builds cleanly, all tests passing. GTM?

koeninger commented:
LGTM  

Thanks for following up on this.

zsxwing commented:
LGTM. Merging to master. Thanks @JasonMWhite and @koeninger
